File: disha-android/app/src/main/java/com/example/gemmachat/core/Summary.kt
package com.example.gemmachat.core

/** Coordinator disaster summary — Kotlin port of disha/core/summary.py. Counts in code; Gemma phrases. */
object Summary {

    private val rank = mapOf("critical" to 0, "high" to 1, "moderate" to 2, "low" to 3)

    data class TopCase(val id: String, val loc: String, val reason: String, val priority: String)
    data class ShelterPressure(val name: String, val occupancy: Int, val capacity: Int, val pressure: Double)
    data class Stats(
        val totalSos: Int, val newSos: Int,
        val critical: Int, val high: Int, val moderate: Int, val low: Int,
        val top5: List<TopCase>, val shortages: Map<String, Int>,
        val shelterPressure: List<ShelterPressure>, val blockedRoads: List<String>,
    )

    fun computeStats(
        sos: List<SosReport>, triage: List<TriageResult>,
        shelters: List<Shelter> = emptyList(), blockedRoads: List<String> = emptyList(),
        newSince: Int = 0,
    ): Stats {
        val sosById = sos.associateBy { it.msgId }
        val ordered = triage.sortedWith(compareBy({ rank[it.priority] }, { -it.urgencyScore })).take(5)
        val top5 = ordered.map { t ->
            val s = sosById[t.msgId]
            val loc = if (s?.lat != null) "%.3f,%.3f".format(s.lat, s.lon) else "loc?"
            TopCase(t.msgId.take(8), loc, t.rationale, t.priority)
        }
        val shortages = HashMap<String, Int>()
        for (t in triage) for (sig in t.riskSignals)
            if (sig == "no_food_water" || sig == "medication_needed")
                shortages[sig] = (shortages[sig] ?: 0) + 1
        return Stats(
            totalSos = sos.size, newSos = newSince,
            critical = triage.count { it.priority == "critical" },
            high = triage.count { it.priority == "high" },
            moderate = triage.count { it.priority == "moderate" },
            low = triage.count { it.priority == "low" },
            top5 = top5, shortages = shortages,
            shelterPressure = shelters.map {
                ShelterPressure(it.name, it.occupancy, it.capacity,
                    Math.round(it.capacityPressure * 100) / 100.0)
            },
            blockedRoads = blockedRoads,
        )
    }

    fun deterministicBriefing(st: Stats): String {
        val sb = StringBuilder()
        sb.append("1) Situation: ${st.totalSos} SOS total (${st.newSos} new).\n")
        sb.append("2) Critical: ${st.critical} | High: ${st.high} | Moderate: ${st.moderate} | Low: ${st.low}.\n")
        sb.append("3) Top cases:\n")
        for (c in st.top5) sb.append("   - ${c.id} · ${c.loc} · ${c.priority} · ${c.reason}\n")
        val sh = st.shortages.entries.joinToString(", ") { "${it.key}×${it.value}" }.ifEmpty { "none reported" }
        sb.append("4) Shortages: $sh.\n")
        if (st.shelterPressure.isNotEmpty()) {
            sb.append("5) Shelter pressure: " +
                st.shelterPressure.joinToString(", ") { "${it.name} ${(it.pressure * 100).toInt()}%" } + ".\n")
        } else sb.append("5) Shelter pressure: n/a.\n")
        sb.append("6) Blocked roads/areas: ${st.blockedRoads.joinToString(", ").ifEmpty { "none reported" }}.\n")
        sb.append("7) Recommended focus: ${if (st.critical > 0) "critical cases first" else "high-priority cases"}.")
        return sb.toString()
    }

    data class Briefing(val briefing: String, val stats: Stats, val producedBy: String)

    fun disasterSummary(
        sos: List<SosReport>, triage: List<TriageResult>, shelters: List<Shelter> = emptyList(),
        blockedRoads: List<String> = emptyList(), newSince: Int = 0, gemma: LlmEngine? = null,
    ): Briefing {
        val st = computeStats(sos, triage, shelters, blockedRoads, newSince)
        if (gemma == null) return Briefing(deterministicBriefing(st), st, "deterministic")
        val user = "NUMBERS + CASES (use only these):\n" + statsToJson(st)
        val briefing = gemma.generate(Prompts.SUMMARY_SYSTEM, user, temperature = 0.3, maxTokens = 350)
        return Briefing(briefing, st, "gemma")
    }

    private fun statsToJson(st: Stats): String = buildString {
        append("{\"total\":${st.totalSos},\"new\":${st.newSos},\"critical\":${st.critical},")
        append("\"high\":${st.high},\"moderate\":${st.moderate},\"low\":${st.low},\"top5\":[")
        append(st.top5.joinToString(",") { "{\"id\":\"${it.id}\",\"loc\":\"${it.loc}\",\"pri\":\"${it.priority}\",\"reason\":\"${it.reason}\"}" })
        append("],\"shortages\":{")
        append(st.shortages.entries.joinToString(",") { "\"${it.key}\":${it.value}" })
        append("},\"blocked\":[${st.blockedRoads.joinToString(",") { "\"$it\"" }}]}")
    }
}

File: disha/core/summary.py
"""Disha coordinator disaster summary (mirrors GEMMA_INTEGRATION.md §6).

Counts are computed HERE (deterministically); Gemma only phrases them, so numbers
can never be hallucinated.
"""
from __future__ import annotations

import json
from collections import Counter
from typing import Optional

from .models import SOSReport, TriageResult, Shelter


def compute_stats(sos: list[SOSReport], triage: list[TriageResult],
                  shelters: Optional[list[Shelter]] = None,
                  blocked_roads: Optional[list[str]] = None,
                  new_since: int = 0) -> dict:
    by_id = {t.msg_id: t for t in triage}
    tiers = Counter(t.priority for t in triage)
    # Top 5 by (tier, urgency)
    rank = {"critical": 0, "high": 1, "moderate": 2, "low": 3}
    ordered = sorted(triage, key=lambda t: (rank[t.priority], -t.urgency_score))[:5]
    top5 = []
    sos_by_id = {s.msg_id: s for s in sos}
    for t in ordered:
        s = sos_by_id.get(t.msg_id)
        loc = (f"{s.lat:.3f},{s.lon:.3f}" if s and s.lat is not None else "loc?")
        top5.append({"id": t.msg_id[:8], "loc": loc, "reason": t.rationale,
                     "priority": t.priority})

    # Resource shortages from flags/signals
    shortage = Counter()
    for t in triage:
        for sig in t.risk_signals:
            if sig in ("no_food_water", "medication_needed"):
                shortage[sig] += 1

    shelter_pressure = []
    if shelters:
        for sh in shelters:
            shelter_pressure.append({"name": sh.name, "occupancy": sh.occupancy,
                                     "capacity": sh.capacity,
                                     "pressure": round(sh.capacity_pressure, 2)})

    return {
        "total_sos": len(sos),
        "new_sos": new_since,
        "critical": tiers.get("critical", 0),
        "high": tiers.get("high", 0),
        "moderate": tiers.get("moderate", 0),
        "low": tiers.get("low", 0),
        "top5": top5,
        "shortages": dict(shortage),
        "shelter_pressure": shelter_pressure,
        "blocked_roads": blocked_roads or [],
    }


def _deterministic_briefing(st: dict) -> str:
    lines = [
        f"1) Situation: {st['total_sos']} SOS total ({st['new_sos']} new).",
        f"2) Critical: {st['critical']} | High: {st['high']} "
        f"| Moderate: {st['moderate']} | Low: {st['low']}.",
        "3) Top cases:",
    ]
    for c in st["top5"]:
        lines.append(f"   - {c['id']} · {c['loc']} · {c['priority']} · {c['reason']}")
    sh = ", ".join(f"{k}×{v}" for k, v in st["shortages"].items()) or "none reported"
    lines.append(f"4) Shortages: {sh}.")
    if st["shelter_pressure"]:
        sp = ", ".join(f"{s['name']} {int(s['pressure']*100)}%" for s in st["shelter_pressure"])
        lines.append(f"5) Shelter pressure: {sp}.")
    else:
        lines.append("5) Shelter pressure: n/a.")
    br = ", ".join(st["blocked_roads"]) or "none reported"
    lines.append(f"6) Blocked roads/areas: {br}.")
    focus = "critical cases first" if st["critical"] else "high-priority cases"
    lines.append(f"7) Recommended focus: {focus}.")
    return "\n".join(lines)


def disaster_summary(sos, triage, shelters=None, blocked_roads=None,
                     new_since=0, gemma=None) -> dict:
    st = compute_stats(sos, triage, shelters, blocked_roads, new_since)
    if gemma is None:
        return {"briefing": _deterministic_briefing(st), "stats": st, "produced_by": "deterministic"}

    from .prompts import SUMMARY_SYSTEM
    user = "NUMBERS + CASES (use only these):\n" + json.dumps(st, ensure_ascii=False, indent=2)
    briefing = gemma.generate(SUMMARY_SYSTEM, user, temperature=0.3, max_tokens=350)
    return {"briefing": briefing, "stats": st, "produced_by": "gemma"}
